<a href="https://colab.research.google.com/github/sdiaz8/sdiaz8/blob/main/Ensemble%20Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#import libraries load iris dataset
import numpy as np
import pandas as pd
import seaborn as sns 
import statsmodels.api as sm
from sklearn.datasets import load_iris
iris = load_iris()

In [ ]:
#create dataframe and showcase features
df = pd.DataFrame(iris.data)
df.columns = iris.feature_names
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [ ]:
#Split Dataset into train and test (30%)
from sklearn.model_selection import train_test_split
y = iris.target
x_train,x_test,y_train,y_test=train_test_split(df,y,test_size=0.3)

In [ ]:
#Create a Gaussian Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=2000, max_depth=2)

In [ ]:
#installing rfpimp
!pip install rfpimp

In [ ]:
#Feature Importance
from sklearn.metrics import r2_score
from rfpimp import permutation_importances

def r2(clf, x_train, y_train):
    return r2_score(y_train, clf.predict(x_train))

perm_imp_rfpimp = permutation_importances(clf, x_train, y_train, r2)
print(perm_imp_rfpimp)

                   Importance
Feature                      
petal width (cm)     0.711028
petal length (cm)    0.217662
sepal length (cm)    0.000000
sepal width (cm)     0.000000


Pedal width is the feature with the most importance.

In [ ]:
#accuracy score using skikitlearn
from sklearn.metrics import accuracy_score
clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
accuracy_score(y_test, y_pred)

0.9333333333333333

An accuracy of 93.3% will be achieved on the test set with bag_clf.

In [ ]:
#Use the Gradient Boosting Algorithm to fit and predict data, and check for accuracy
from sklearn.ensemble import GradientBoostingClassifier
gbc = GradientBoostingClassifier(n_estimators=2000, max_depth=2)
gbc.fit(x_train,y_train)
y_pred2=gbc.predict(x_test)
accuracy_score(y_test, y_pred2)

0.9555555555555556

In [ ]:
#Feature Importance (Gradient)
def r2(gbc, x_train, y_train):
    return r2_score(y_train, gbc.predict(x_train))

perm_imp_rfpimp = permutation_importances(gbc, x_train, y_train, r2)
print(perm_imp_rfpimp)

                   Importance
Feature                      
petal width (cm)     0.725539
petal length (cm)    0.565920
sepal length (cm)    0.014511
sepal width (cm)     0.014511


The Gradient Boosting Model performed better, the only reservation that I have is the model may be overfitted.  It would be susceptible to high variance and low bias if additional data is presented.